lib/class-wp-block.php

```diff
@@ -0,0 +1,234 @@
+	 * Block context values.
+	 *
+	 * @var array
+	 */
+	public $context = [];
+
+	/**
+	 * All available context of the current hierarchy.
+	 *
+	 * @var array
+	 * @access protected
+	 */
+	protected $available_context;
+
+	/**
+	 * Block attribute values.
+	 *
+	 * @var array
+	 */
+	public $attributes = [];
+
+	/**
+	 * List of inner blocks (of this same class)
+	 *
+	 * @var WP_Block[]
+	 */
+	public $inner_blocks = [];
+
+	/**
+	 * Resultant HTML from inside block comment delimiters after removing inner
+	 * blocks.
+	 *
+	 * @example "...Just <!-- wp:test /--> testing..." -> "Just testing..."
+	 *
+	 * @var string
+	 */
+	public $inner_html = '';
+
+	/**
+	 * List of string fragments and null markers where inner blocks were found
+	 *
+	 * @example array(
+	 *   'inner_html'    => 'BeforeInnerAfter',
+	 *   'inner_blocks'  => array( block, block ),
+	 *   'inner_content' => array( 'Before', null, 'Inner', null, 'After' ),
+	 * )
+	 *
+	 * @var array
+	 */
+	public $inner_content = [];
+
+	/**
+	 * Constructor.
+	 *
+	 * Populates object properties from the provided block instance argument.
+	 *
+	 * The given array of context values will not necessarily be available on
+	 * the instance itself, but is treated as the full set of values provided by
+	 * the block's ancestry. This is assigned to the private `available_context`
+	 * property. Only values which are configured to consumed by the block via
+	 * its registered type will be assigned to the block's `context` property.
+	 *
+	 * @param array                  $block    Array of parsed block properties.
+	 * @param array                  $context  Optional array of ancestry context values.
+	 * @param WP_Block_Type_Registry $registry Optional block type registry.
+	 */
+	public function __construct( $block, $context = [], $registry = null ) {
+		$this->name = $block['blockName'];
+
+		if ( is_null( $registry ) ) {
+			$registry = WP_Block_Type_Registry::get_instance();
+		}
+
+		$this->block_type = $registry->get_registered( $this->name );
+
+		if ( ! empty( $block['attrs'] ) ) {
+			$this->attributes = $block['attrs'];
+		}
+
+		if ( ! is_null( $this->block_type ) ) {
+			$this->attributes = $this->block_type->prepare_attributes_for_render( $this->attributes );
+		}
+
+		$this->available_context = $context;
+
+		if ( ! empty( $this->block_type->context ) ) {
+			foreach ( $this->block_type->context as $context_name ) {
+				if ( array_key_exists( $context_name, $this->available_context ) ) {
+					$this->context[ $context_name ] = $this->available_context[ $context_name ];
+				}
+			}
+		}
+
+		if ( ! empty( $block['innerBlocks'] ) ) {
+			$child_context = $this->available_context;
+
+			/* phpcs:disable WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase */
+			if ( ! empty( $this->block_type->providesContext ) ) {
+				foreach ( $this->block_type->providesContext as $context_name => $attribute_name ) {
+					if ( isset( $this->attributes[ $attribute_name ] ) ) {
```

Comment:
> Should a `null` attribute override parent context? If so, this should be `array_key_exists()` instead of `isset()`.
> 
> https://devdocs.io/php/function.array-key-exists
> 
> > isset() does not return TRUE for array keys that correspond to a NULL value, while array_key_exists() does.

Good point! I'll make the change, and include a test case.

Reply:
> Should a `null` attribute override parent context? If so, this should be `array_key_exists()` instead of `isset()`.

Fixed in f11c515.